### PR TITLE
Add custom forwarding routes feature [19249]

### PR DESCRIPTION
**Pull request by juanlofer-eprosima**
Pending tasks:
- [x] YAML tests
- [x] Revision

### CODE REVIEW

**Review by Tempate — CHANGES_REQUESTED** [review]

File: ddspipe_yaml/src/cpp/YamlReader_features.cpp [hunk 1]
```diff
@@ -59,6 +65,150 @@ participants::XmlHandlerConfiguration YamlReader::get(
...
+/************************
+* Routes Configuration  *
+************************/
+
+template <>
+DDSPIPE_YAML_DllAPI
+void YamlReader::fill(
+        core::RoutesConfiguration& object,
+        const Yaml& yml,
+        const YamlReaderVersion version)
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddspipe_yaml/src/cpp/YamlReader_features.cpp:77]: There is no performance gain for enums, just like it is the case for booleans for example.

File: ddspipe_core/include/ddspipe_core/configuration/RoutesConfiguration.hpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    // CONSTRUCTORS
+    /////////////////////////
+
+    DDSPIPE_CORE_DllAPI RoutesConfiguration() = default;
+
+    /////////////////////////
+    // METHODS
+    /////////////////////////
+
+    DDSPIPE_CORE_DllAPI virtual bool is_valid(
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddspipe_core/include/ddspipe_core/configuration/RoutesConfiguration.hpp:47]: As far as I know, virtual and override keywords are optional in overriding classes. So no, but it does not hurt to have it.

File: ddspipe_core/include/ddspipe_core/configuration/TopicRoutesConfiguration.hpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    // CONSTRUCTORS
+    /////////////////////////
+
+    DDSPIPE_CORE_DllAPI TopicRoutesConfiguration() = default;
+
+    /////////////////////////
+    // METHODS
+    /////////////////////////
+
+    DDSPIPE_CORE_DllAPI virtual bool is_valid(
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddspipe_core/include/ddspipe_core/configuration/TopicRoutesConfiguration.hpp:48]: Same answer as before.

File: ddspipe_core/src/cpp/communication/dds/DdsBridge.cpp [hunk 4]
```diff
@@ -27,53 +27,118 @@ DdsBridge::DdsBridge(
...
+                id,
+                readers[id],
+                std::move(dst_writers),
+                payload_pool,
+                thread_pool);
+        }
+        else
+        {
+            // List of all Participants
+            std::map<ParticipantId, std::shared_ptr<IWriter>> writers_except_one =
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddspipe_core/src/cpp/communication/dds/DdsBridge.cpp:118]: Is it?

[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddspipe_core/src/cpp/communication/dds/DdsBridge.cpp:118]: And incompatible with merging if-else statements into one suggestion.

File: ddspipe_yaml/include/ddspipe_yaml/yaml_configuration_tags.hpp [hunk 5]
```diff
@@ -122,6 +122,12 @@ constexpr const char* TIMESTAMP_MILLISECONDS_TAG("milliseconds"); //! TODO: add
 constexpr const char* TIMESTAMP_MICROSECONDS_TAG("microseconds"); //! TODO: add comment
 constexpr const char* TIMESTAMP_NANOSECONDS_TAG("nanoseconds"); //! TODO: add comment
 
+// Custom routes related tags
+constexpr const char* ROUTES_TAG("routes"); //! Forwarding routes
+constexpr const char* ROUTES_SRC_TAG("src"); //! Route source
```
Comment (Tempate) [inline ddspipe_yaml/include/ddspipe_yaml/yaml_configuration_tags.hpp:127]: I would write "Route's source participant" and "Route's destination participants"

File: ddspipe_core/src/cpp/core/DdsPipe.cpp [hunk 6]
```diff
@@ -343,7 +347,16 @@ void DdsPipe::create_new_bridge_nts_(
 
     try
     {
-        auto new_bridge = std::make_unique<DdsBridge>(topic, participants_database_, payload_pool_, thread_pool_);
+        // Use topic specific forwarding routes if available
+        RoutesConfiguration routes_config = topic_routes_config_().count(topic) !=
```
Comment (Tempate) [inline ddspipe_core/src/cpp/core/DdsPipe.cpp:351]: I think it would be more readable if the one-liner if statement was replaced by a normal if statement.

**Review by Tempate — CHANGES_REQUESTED** [review]

**Review by Tempate — APPROVED** [review]
LGTM

**Pull request merged**